File: 3646.next-special-palindrome-number.py
#
# @lc app=leetcode id=3646 lang=python3
#
# [3646] Next Special Palindrome Number
#

# @lc code=start
class Solution:
    def specialPalindrome(self, n: int) -> int:
        def is_special_palindrome(s):
            counts = {str(i): 0 for i in range(10)}
            for ch in s:
                counts[ch] += 1
            for k in counts:
                if counts[k] != 0 and int(k) != counts[k]:
                    return False
            return True
        
        def generate_next_palindrome(s):
            l = len(s)
            half = (l + 1) // 2
            left = s[:half]
            if l % 2 == 0:
                candidate = left + left[::-1]
            else:
                candidate = left + left[-2::-1]
            if candidate > s:
                return candidate
            # Increment left half and create new palindrome with carry handling
            left_as_num = str(int(left) + 1)
            if len(left_as_num) > half:
                return '1' + '0' * (l - 1) + '1'
            if l % 2 == 0:
                return left_as_num + left_as_num[::-1]
            else:
                return left_as_num + left_as_num[-2::-1]
        
        # Start with the smallest palindrome greater than n
        n_str = str(n)
        candidate = generate_next_palindrome(n_str)
        while not is_special_palindrome(candidate):
            candidate = generate_next_palindrome(candidate)
        return int(candidate)
# @lc code=end

File: 3646.next-special-palindrome-number.go
#
# @lc app=leetcode id=3646 lang=golang
#
# [3646] Next Special Palindrome Number
#
# @lc code=start
import (
    "strconv"
)

func specialPalindrome(n int64) int64 {
    // Helper function to check if a number is special
    hasValidDigitCounts := func(num int64) bool {
        counts := make(map[rune]int)
        s := strconv.FormatInt(num, 10)
        for _, digit := range s {
            counts[digit]++
        }
        for digit, count := range counts {
            if int(digit-'0') != count { 
                return false
            }
        }
        return true
    }

    // Generate next palindrome candidate greater than n
    generateNextPalindrome := func(num int64) int64 {
        s := strconv.FormatInt(num, 10)
        length := len(s)
        halfLength := length / 2
        incremented := false
        var candidate string
        if length%2 == 0 {
            firstHalf, _ := strconv.ParseInt(s[:halfLength], 10, 64)
            firstHalf++
            incremented = true
            firstHalfStr := strconv.FormatInt(firstHalf, 10)
            candidate = firstHalfStr + reverseString(firstHalfStr)
        } else {
            middleIndex := halfLength
            firstHalf, _ := strconv.ParseInt(s[:middleIndex+1], 10, 64)
            firstHalf++
            incremented = true
            firstHalfStr := strconv.FormatInt(firstHalf, 10)
            candidate = firstHalfStr + reverseString(firstHalfStr[:middleIndex])
        }
        result, _ := strconv.ParseInt(candidate, 10, 64)
        return result
    }
    	// Reverse string utility function	reverseString := func(s string) string {		runes := []rune(s)		for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {			runes[i], runes[j] = runes[j], runes[i]		}		return string(runes)	}		candidate := n + 1	for {	candidate = generateNextPalindrome(candidate) // generate next palindrome	if hasValidDigitCounts(candidate){	return candidate	}	candidate++ // ensure we are always moving forward	}	}